lighthouse-core/report/v2/renderer/report-features.js

```diff
@@ -0,0 +1,326 @@
+    }
+  }
+
+  closeExportDropdown() {
+    this.exportButton.classList.remove('active');
+  }
+
+  /**
+   * Click handler for export button.
+   * @param {!Event} e
+   */
+  onExportButtonClick(e) {
+    e.preventDefault();
+    const el = /** @type {!Element} */ (e.target);
+    el.classList.toggle('active');
+    this._document.addEventListener('keydown', this.onKeyDown);
+  }
+
+  /**
+   * Resets the state of page before capturing the page for export.
+   * When the user opens the exported HTML page, certain UI elements should
+   * be in their open state (not opened) and the templates should be refreshly
+   * stamped.
+   */
+  _resetUIState() {
+    this.logger.hide();
+    this.closeExportDropdown();
+    this._dom.findAll('template[data-stamped]', this._document).forEach(t => {
+      t.removeAttribute('data-stamped');
+    });
+  }
+
+  /**
+   * Handler for "export as" button.
+   * @param {!Event} e
+   */
+  onExport(e) {
+    e.preventDefault();
+
+    const el = /** @type {!Element} */ (e.target);
+
+    if (!el.hasAttribute('data-action')) {
+      return;
+    }
+
+    switch (el.getAttribute('data-action')) {
+      case 'copy':
+        this.onCopyButtonClick();
+        break;
+      case 'open-viewer':
+        this.sendJSONReport();
+        break;
+      case 'print':
+        this.expandAllDetails();
+        self.print();
+        break;
+      case 'save-json': {
+        const jsonStr = JSON.stringify(this.json, null, 2);
+        this._saveFile(new Blob([jsonStr], {type: 'application/json'}));
+        break;
+      }
+      case 'save-html': {
+        this._resetUIState();
+
+        const htmlStr = this._document.documentElement.outerHTML;
+        try {
+          this._saveFile(new Blob([htmlStr], {type: 'text/html'}));
+        } catch (/** @type {!Error} */ e) {
+          this.logger.error('Could not export as HTML. ' + e.message);
+        }
+        break;
+      }
+    }
+
+    this.closeExportDropdown();
+    this._document.removeEventListener('keydown', this.onKeyDown);
+  }
+
+  /**
+   * Keydown handler for the document.
+   * @param {!Event} e
+   */
+  onKeyDown(e) {
+    if (e.keyCode === 27) { // ESC
+      this.closeExportDropdown();
+    }
+  }
+
+  /**
+   * Opens a new tab to the online viewer and sends the local page's JSON results
+   * to the online viewer using postMessage.
+   */
+  sendJSONReport() {
+    const VIEWER_ORIGIN = 'https://googlechrome.github.io';
+    const VIEWER_URL = `${VIEWER_ORIGIN}/lighthouse/viewer/`;
+
+    // Chrome doesn't allow us to immediately postMessage to a popup right
+    // after it's created. Normally, we could also listen for the popup window's
+    // load event, however it is cross-domain and won't fire. Instead, listen
+    // for a message from the target app saying "I'm open".
+    self.addEventListener('message', function msgHandler(/** @type {!Event} */ evt) {
+      const e = /** @type {!MessageEvent} */ (evt);
```

Comment:
`MessageEvent` is a [templated type](https://github.com/google/closure-compiler/blob/0a2b4d12fa3f30cfe4d9a184f3c0cc3e349313d4/externs/browser/html5.js#L2168-L2174), so you can type the `data` property on it in the type declaration. e.g. `/** @type {!MessageEvent<{opened: boolean}>} */`